gtsam/geometry/Line3.h

```diff
@@ -0,0 +1,178 @@
+ * @author  Frank Dellaert
+ */
+// \callgraph
+
+#pragma once
+
+#include <gtsam/base/concepts.h>
+#include <gtsam/geometry/Rot3.h>
+#include <gtsam/geometry/SO3.h>
+#include <gtsam/nonlinear/expressions.h>
+#include <gtsam/slam/expressions.h>
+
+#include <boost/random.hpp>
+#include <cstdlib>
+#include <ctime>
+
+namespace gtsam{
+
+/**
+ * A 3D line (R,a,b) : (Rot3,Scalar,Scalar)
+ * @addtogroup geometry
+ * \nosubgrouping
+ */
+class Line3 {
+private:
+    Rot3 R_;    // Rotation of line about x and y in world frame
+    double a_, b_;  // Intersection of line with x-y world plane
+
+public:
+    enum {dimension = 4};
+
+    /** Default constructor is the Z axis **/
+    Line3() :
+            a_(0), b_(0)
+    {}
+
+    /** Constructor:
+     * Parallel to z axis, intersecting x-y plane at (a,b) **/
+    Line3(const double a, const double b):
+            a_(a), b_(b)
+    {}
+
+    /** Constructor for general line from (R, a, b) **/
+    Line3(const Rot3 &R, const double a, const double b) :
+            R_(R), a_(a), b_(b)
+    {}
+
+    /** Sample two points lying on the line
+     * @return Vector6 where first 3 elements are one point, and
+     * next 3 elements are another point **/
+    Vector6 points();
+
+    /**
+     * The retract method maps from the tangent space back to the manifold.
+     * The tangent space for the rotation of a line is only two dimensional -
+     * rotation about x and y
+     * @param v: increment in tangent space
+     * @param H: Jacobian of retraction with respect to the increment
+     * @return: resulting line after adding the increment and mapping to the manifold
+     */
+    Line3 retract(const Vector4 &v, OptionalJacobian<4,4> H = boost::none) const;
+
+    /**
+     * The localCoordinates method is the inverse of retract and finds the difference
+     * between two lines in the tangent space.
+     * @param q Line3 on manifold
+     * @param H OptionalJacobian of localCoordinates with respect to line
+     * @return Vector4 difference in the tangent space
+     */
+    Vector4 localCoordinates(const Line3 &q, OptionalJacobian<4,4> H = boost::none) const;
+
+    /**
+     * Rotation of line accessor
+     * @return Rot3
+     */
+    Rot3 R() const
+    {
+        return R_;
+    }
+
+    /**
+     * Accessor for a, b
+     * @return Vector2(a, b)
+     */
+    Vector2 V() const
+    {
+        return Vector2(a_, b_);
+    }
+
+    /**
+     * Print R, a, b
+     * @param s: optional starting string
+     */
+    void print(const std::string &s = "") const;
+
+    /**
+     * Check if two lines are equal
+     * @param l2 - line to be compared
+     * @param tol : optional tolerance
+     * @return boolean - true if lines are equal
+     */
+    bool equals(const Line3 &l2, double tol = 10e-9) const;
+
+    /**
+     * Projecting a line to an image plane
+     * @param Dline: OptionalJacobian of projected line with respect to this line
+     * @return Point3 - projected line in image plane
```

Comment:
I don't understand. Why is this a Point3? Properly comment :-)

Reply:
It is the 2D line in the image plane in homogenous coordinates. It could be a Vec3, Point3 or Unit3. Could you tell me which one is to be preferred and why?

Reply:
It should be between Vector3 and Unit3, as Point3 really has the "I'm a point in 3D" name :-) Unit3 is in a sense more correct as it has the correct manifold dimension, but it will incur a little bit more computation (as it normalizes). Then again, the sphere and P^3 (the projective space of lines are *not* the same. Still, Unit3 is a bit like a Quaternion representation for 2D lines, so I think it's the best we have. 

Part of the comment is also about documentation, though: say "2D line in the image plane in homogenous coordinates. We use Unit3 as the return type as that is a manifold of the right dimension."

Reply:
This might have some effect on my reconstruction. The returned optionalJacobian would change from having a shape <3,4> to <2,4>. I will try it out.

Reply:
Indeed! And very much correctly so :-) 2D lines *are* only 2-dimensional.